drivers/net/ethernet/intel/idpf/idpf_eth_idc.c

```diff
@@ -0,0 +1,453 @@
+/**
+ * idpf_eth_idc_driver_register - Register ethernet auxiliary driver
+ * @void: void
+ *
+ * Returns 0 on success, negative on failure
+ */
+int idpf_eth_idc_driver_register(void)
+{
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+
+	eth_idc_drv = idpf_eth_idc_get_driver();
+	return auxiliary_driver_register(&eth_idc_drv->adrv);
+}
+
+/**
+ * idpf_eth_idc_driver_unregister - unregister ethernet driver
+ * @void: void
+ */
+void idpf_eth_idc_driver_unregister(void)
+{
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+
+	eth_idc_drv = idpf_eth_idc_get_driver();
+	auxiliary_driver_unregister(&eth_idc_drv->adrv);
+}
+
+/**
+ * idpf_eth_idc_init_device_params - Initialize device params
+ * @adapter: Idpf private structure
+ * @eth_dev: Ethernet dev info
+ */
+static void
+idpf_eth_idc_init_device_params(struct idpf_adapter *adapter,
+				struct idpf_eth_idc_auxiliary_dev *eth_dev)
+{
+	/* crc info */
+	eth_dev->eth_info.caps.crc_enable = adapter->crc_enable;
+
+	/* Queue Info */
+	eth_dev->eth_info.caps.q_info.max_rxq =
+		le16_to_cpu(adapter->caps.max_rx_q);
+	eth_dev->eth_info.caps.q_info.max_txq =
+		le16_to_cpu(adapter->caps.max_tx_q);
+	eth_dev->eth_info.caps.q_info.max_bufq =
+		le16_to_cpu(adapter->caps.max_rx_bufq);
+	eth_dev->eth_info.caps.q_info.max_complq =
+		le16_to_cpu(adapter->caps.max_tx_complq);
+
+	/* Caps */
+	eth_dev->eth_info.caps.csum_caps = adapter->caps.csum_caps;
+	eth_dev->eth_info.caps.seg_caps = adapter->caps.seg_caps;
+	eth_dev->eth_info.caps.hsplit_caps =
+		adapter->caps.hsplit_caps;
+	eth_dev->eth_info.caps.rsc_caps = adapter->caps.rsc_caps;
+	eth_dev->eth_info.caps.rss_caps = adapter->caps.rss_caps;
+	eth_dev->eth_info.caps.other_caps = adapter->caps.other_caps;
+	eth_dev->eth_info.caps.min_sso_packet_len =
+		adapter->caps.min_sso_packet_len;
+	eth_dev->eth_info.caps.max_sg_bufs_per_tx_pkt =
+		adapter->caps.max_sg_bufs_per_tx_pkt;
+}
+
+/**
+ * idpf_eth_idc_device_alloc - allocate auxiliary device
+ * @adapter: Idpf private structure
+ * @vport_type: Vport type
+ *
+ * Returns allocated structure
+ */
+static struct idpf_eth_idc_auxiliary_dev *
+idpf_eth_idc_device_alloc(struct idpf_adapter *adapter,
+			  enum idpf_vport_type vport_type)
+{
+	struct idpf_eth_idc_auxiliary_dev *new_eth_dev;
+	struct idpf_eth_idc_dev_info *dev_info;
+	enum idpf_queue_type queue_type;
+	u16 i, err;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
+		/* Check pre-allocated one first */
+		if (adapter->adevs[i])
+			continue;
+
+		adapter->adevs[i] =
+			kzalloc(sizeof(struct idpf_eth_idc_auxiliary_dev),
+				GFP_KERNEL);
+		new_eth_dev = adapter->adevs[i];
+		if (!new_eth_dev)
+			goto alloc_exit;
+
+		new_eth_dev->adev.id = ida_alloc(&idpf_eth_idc_ida, GFP_KERNEL);
+		new_eth_dev->eth_info.eth_shared = &adapter->eth_shared;
+		new_eth_dev->eth_info.idx = i;
+
+		/* Set vport type */
+		new_eth_dev->eth_info.vport_type = vport_type;
+
+		if (vport_type == IDPF_DEFAULT_VPORT)
+			queue_type = IDPF_QUEUE_DEFAULT_VPORT;
```

Comment:
why do we need another enum? can we reuse the IDPF_DEFAULT_VPORT to differentiate the number of queues to be allocated?

Reply:
The function is dealing with the queues not vports.